File: fe/ClinicManager/src/models/payment.ts
import { Voucher } from "./voucher";

export interface Payment {
    id: number;
    method: PaymentMethod;
    amount: number;
    voucher: Voucher | null;
}

export interface NewPayment {
    method: PaymentMethod | null;
    amount: number;
    voucher?: Voucher | null; /* only if paid by Voucher attach it */
}

export enum PaymentMethod {
    CASH = "CASH",
    CARD = "CARD",
    BLIK = "BLIK",
    TRANSFER = "TRANSFER",
    VOUCHER = "VOUCHER"
}

export enum PaymentStatus {
    PAID = "PAID",
    PARTIAL = "PARTIAL",
    UNPAID = "UNPAID"
}

File: be/ClinicManager/src/main/java/com/clinic/clinicmanager/model/Payment.java
package com.clinic.clinicmanager.model;

import com.clinic.clinicmanager.model.constants.PaymentMethod;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;

@Entity
@Getter
@Setter
@SuperBuilder
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "payment")
public class Payment {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PaymentMethod method;

    @Column(nullable = false)
    private Double amount;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "voucher_id")
    private Voucher voucher; // if paid by voucher attach it

    @PrePersist
    @PreUpdate
    private void validatePayment() {
        if (method == PaymentMethod.VOUCHER && voucher == null) {
            throw new IllegalStateException("Voucher must be assigned for VOUCHER payment method.");
        }
        if (method != PaymentMethod.VOUCHER && voucher != null) {
            throw new IllegalStateException("Voucher should only be assigned when payment method is VOUCHER.");
        }
    }
}
